### DISCUSSION TITLE
Creating a new user directly through postgres - failed login

### DISCUSSION THREAD

**Discussion by rotemrevivo91** [supabase/supabase#5248] [Questions] [answered]
I'm creating a new user directly from postgres at docker-compose build time by adding a new sql file to `supabase/docker/db/init`:
```sql
insert into auth.users (
  id, instance_id, role, email, 
  raw_app_meta_data, raw_user_meta_data, is_super_admin,
  encrypted_password,
  created_at, updated_at, last_sign_in_at, confirmed_at
)
values (
  gen_random_uuid(), '00000000-0000-0000-0000-000000000000', 'authenticated', 'REPLACE_ADMIN_EMAIL', 
  '{"provider":"email","providers":["email"]}', '{}', false, 
  crypt('REPLACE_ADMIN_PASSWORD', gen_salt('bf')),
  NOW(), NOW(), NOW(), NOW()
);
``` 
It succeeds and i can see the new user in the auth tab and also in the auth.users table. Now when i'm trying to login i get an error 400: 
```js
{
  "error": "invalid_grant"
  "error_description": "Invalid login credentials"
}
```

I'm not sure if i'm missing a step or if the encryption of the password is done through supabase in a different manner. But, the following did work:
```sql
SELECT id
FROM auth.users
WHERE encrypted_password = crypt('mypassword', encrypted_password);
```
(Thanks to ftonato: https://github.com/supabase/supabase/discussions/627)

I would appreciate any help. Thanks!

**Comment by GaryAustin1** [comment]
I've used this:
```
 UPDATE auth.users SET
  encrypted_password =
  crypt(new_plain_password, gen_salt('bf'))
  WHERE id = auth.uid();
```
So I think you are good on encryption...

My user table has an "aud" column set to "authenticated". not sure what aud stands for.

**Reply by rotemrevivo91** [reply]
I tried adding it as well but still nothing changed

**Comment by rotemrevivo91** [comment]
I tried a couple more things but nothing really worked. I was able to insert a row to `auth.identities` even though for some reason its definition is missing from `supabase/docker/volumes/db/init/01-auth-schema.sql`. I found the definition in the `supbase/cli` repo:
```sql
-- auth.identities definitions
CREATE TABLE auth.identities (
    id text NOT NULL,
    user_id uuid NOT NULL,
    identity_data jsonb NOT NULL,
    provider text NOT NULL,
    last_sign_in_at timestamp with time zone,
    created_at timestamp with time zone,
    updated_at timestamp with time zone
);
COMMENT ON TABLE auth.identities IS 'Auth: Stores identities associated to a user.';
CREATE INDEX identities_user_id_idx ON auth.identities USING btree (user_id);
.
.
ALTER TABLE auth.identities OWNER TO supabase_auth_admin;
```

Then in the additional file i added to the init folder:
```sql
do $$
  declare
    uid uuid = gen_random_uuid();
    now timestamp = NOW();
    idata jsonb = '{"sub":"' || uid || '"}';

  begin
    insert into auth.users (
      id, instance_id, role, email, 
      raw_app_meta_data, raw_user_meta_data, is_super_admin,
      encrypted_password,
      created_at, updated_at, last_sign_in_at, confirmed_at
    )
    values (
      uid, '00000000-0000-0000-0000-000000000000', 'authenticated', 'REPLACE_ADMIN_EMAIL', 
      '{"provider":"email","providers":["email"]}', '{}', false, 
      crypt('REPLACE_ADMIN_PASSWORD', gen_salt('bf')),
      now, now, now, now
    );

    insert into auth.identities (
      id, provider, user_id, identity_data, last_sign_in_at, created_at, updated_at
    )
    values (uid, 'email', uid, idata, now, now, now);
end;
$$ language plpgsql;
```

A new row is generated for both auth.users and auth.identites. The user is visible in the auth tab as well.

**Comment by rotemrevivo91** [comment, marked as answer]
I think that for now i have a different probably much better and simpler solution that i can't believe i didn't think of before for some reason...
So i'll just create a bash script that will run after the containers are up that will make  a post call to gotrue for signing up a new user and gotrue will take care of all creating all the necessary rows and permissions.
```sh
curl --silent --output /dev/null 'https://localhost:8000/auth/v1/signup' \
-H "apikey: ..." \
-d '{"email":"...", "password":"..."}' 
```

**Reply by naxprmn** [reply]
hi rotem, currently Im also struggling for generating user at supabase, would you kindly post full tutorial on this case?  what I want to achieve is Im able to import my username and password from csv. thank you in advance

maybe if you have free time you can answer here https://stackoverflow.com/questions/71552961/how-to-import-csv-generate-user-table-at-supabase thanks!

**Reply by rotemrevivo91** [reply]
Are you trying to achieve that right after the docker containers are up? If so, I have done it as a shell script:
```sh
CURL_RES=22
count=1
echo "Creating user ${ADMIN_EMAIL}"
while [[ $CURL_RES = 22 && count -le 5 ]] ; do
  echo "Try ${count}/5"
  sleep 5
  CURL_RES=$(curl --silent "${API_URL}/auth/v1/signup" -H "apikey: ${ANON_KEY}" -d '{"email":"'"${ADMIN_EMAIL}"'", "password":"'"${ADMIN_PASSWORD}"'"}')
  let "count+=1"
done;

if [[ $CURL_RES = 22 ]]; then
  echo "Failed creating ${ADMIN_EMAIL}..Removing containers."
  eval "sh stop.sh"
else
  echo "Successfully created ${ADMIN_EMAIL}"
fi;
```
The loop and the if statement are there just to make sure that the containers are already up, and if not give it some time to go up before succeeding or failing .

**Reply by rotemrevivo91** [reply]
It really depends what you want to achieve. I needed a simple way to run my containers and init them already with a user for my use case. That's why writing a small script on a linux server was the way I chose to do it.
In general, you can access the signup url of your deployment using the anon key and pass it an object with the e-mail and password. 
According to your stackoverflow post you have a problem with encrypted passwords so I'm not sure if i can offer any help.

**Reply by naxprmn** [reply]
yeah my problem are with encrypted password, anyway thanks for your thought, I really appreciated it. and sorry for the long time reply. Thanks again rotem :D

**Comment by msdrigg** [comment]
I found this trying to answer the question "How are the users themselves actually stored in the db?", so I was curious why this didn't work for you. To answer your question. I do believe that `crypt` for encrypting the password is the incorrect method. Supabase uses goTrue which hashes passwords using bcrypt, not the postgres `crypt` function. See the `hashPassword` function here https://github.com/supabase/gotrue/blob/master/models/user.go for more details.

**Reply by msdrigg** [reply]
I know this discussion has been resolved, but leaving here in case others get here from google.

**Comment by raimille1** [comment]
I know some people might be interested in this. We have a simple seed.sql file, so we needed to solve it the SQL way. Most answers were almost there, but the reason the login was throwing 500s is because some "NULLABLE" fields in `auth.users` are not really nullable on the Go world that reads them. All you need to do is specify that all the _token columns be initialized to empty strings rather than NULL. We did the following and it's working:

```
INSERT INTO
  auth.users (
    id,
    instance_id,
    ROLE,
    aud,
    email,
    raw_app_meta_data,
    raw_user_meta_data,
    is_super_admin,
    encrypted_password,
    created_at,
    updated_at,
    last_sign_in_at,
    email_confirmed_at,
    confirmation_sent_at,
    confirmation_token,
    recovery_token,
    email_change_token_new,
    email_change
  )
VALUES
  (
    gen_random_uuid(),
    '00000000-0000-0000-0000-000000000000',
    'authenticated',
    'authenticated',
    'dev@email.io',
    '{"provider":"email","providers":["email"]}',
    '{}',
    FALSE,
    crypt('Pa55word!', gen_salt('bf')),
    NOW(),
    NOW(),
    NOW(),
    NOW(),
    NOW(),
    '',
    '',
    '',
    ''
  );

INSERT INTO
  auth.identities (
    id,
    provider,
    user_id,
    identity_data,
    last_sign_in_at,
    created_at,
    updated_at
  )
VALUES
  (
    (
      SELECT
        id
      FROM
        auth.users
      WHERE
        email = 'dev@email.io'
    ),
    'email',
    (
      SELECT
        id
      FROM
        auth.users
      WHERE
        email = 'dev@email.io'
    ),
    json_build_object(
      'sub',
      (
        SELECT
          id
        FROM
          auth.users
        WHERE
          email = 'dev@email.io'
      )
    ),
    NOW(),
    NOW(),
    NOW()
  );

```

Hopefully it helps a few!

**Reply by redarccoder** [reply]
Thanks for this reply, really helped me fix my setup : )

**Reply by ozum** [reply]
@raimille1, thanks, this is what I was looking for (`seed.sql`). I modified your query a little. In the modified query, you just put email in one place and can create multiple users simultaneously.

```sql
WITH credentials(mail, pass) AS (
  -- PUT YOUR EMAILS AND PASSWORDS HERE.
  SELECT * FROM (VALUES ('dev_1@email.io', 'passWORD'), ('dev_2@email.io', 'passWORD')) AS users
),
create_user AS (
  INSERT INTO auth.users (id, instance_id, ROLE, aud, email, raw_app_meta_data, raw_user_meta_data, is_super_admin, encrypted_password, created_at, updated_at, last_sign_in_at, email_confirmed_at, confirmation_sent_at, confirmation_token, recovery_token, email_change_token_new, email_change)
    SELECT gen_random_uuid(), '00000000-0000-0000-0000-000000000000', 'authenticated', 'authenticated', mail, '{"provider":"email","providers":["email"]}', '{}', FALSE, crypt(pass, gen_salt('bf')), NOW(), NOW(), NOW(), NOW(), NOW(), '', '', '', '' FROM credentials
  RETURNING id
)
INSERT INTO auth.identities (id, provider, user_id, identity_data, last_sign_in_at, created_at, updated_at)
  SELECT id, 'email', id, json_build_object('sub', id), NOW(), NOW(), NOW() FROM create_user;


```

**Reply by michelcrypt4d4mus** [reply]
> _I know some people might be interested in this. We have a simple seed.sql file, so we needed to solve it the SQL way. Most answers were almost there, but the reason the login was throwing 500s is because some "NULLABLE" fields in `auth.users` are not really nullable on the Go world that reads them._

this no longer works because `auth.identities.provider_id` is no longer allowed to be NULL. Fix was pretty straightforward though:

```sql
INSERT INTO
  auth.users (
    id,
    instance_id,
    ROLE,
    aud,
    email,
    raw_app_meta_data,
    raw_user_meta_data,
    is_super_admin,
    encrypted_password,
    created_at,
    updated_at,
    last_sign_in_at,
    email_confirmed_at,
    confirmation_sent_at,
    confirmation_token,
    recovery_token,
    email_change_token_new,
    email_change
  )
VALUES
  (
    gen_random_uuid(),
    '00000000-0000-0000-0000-000000000000',
    'authenticated',
    'authenticated',
    'dev@bestcompany.ch',
    '{"provider":"email","providers":["email"]}',
    '{}',
    FALSE,
    crypt('passWERDZ', gen_salt('bf')),
    NOW(),
    NOW(),
    NOW(),
    NOW(),
    NOW(),
    '',
    '',
    '',
    ''
  );



INSERT INTO
  auth.identities (
    id,
    provider_id,
    provider,
    user_id,
    identity_data,
    last_sign_in_at,
    created_at,
    updated_at
  )
VALUES
  (
    (SELECT id FROM auth.users WHERE email = 'dev@bestcompany.ch'),
    (SELECT id FROM auth.users WHERE email = 'dev@bestcompany.ch'),
    'email',
    (SELECT id FROM auth.users WHERE email = 'dev@bestcompany.ch'),
    json_build_object('sub', (SELECT id FROM auth.users WHERE email = 'dev@bestcompany.ch')),
    NOW(),
    NOW(),
    NOW()
  );
```

**Reply by Daniel-HyunHo-Shin** [reply]
nice

**Reply by holysoros** [reply]
if you encountered 'no gen_salt' error, change the line `crypt('passWERDZ', gen_salt('bf'))` to `extensions.crypt('Abcd1234$', extensions.gen_salt('bf'))`

**Reply by itsmikesharescode** [reply]
is this still works ?

**Reply by celine-m-s** [reply]
Thank you so much!

**Comment by jcoble** [comment]
looks goood

**Comment by jdgamble555** [comment]
I think this no longer works due to locked schemas.

https://github.com/orgs/supabase/discussions/34270

Can we get official support for this?

J

**Reply by silentworks** [reply]
You can still create new users directly through postgres, this hasn't changed with the locked schemas. Locked schemas restricts you from modifying the schema in a destructive way, inserting data into the `users` table in the `auth` schema is not destructive.

**Reply by jdgamble555** [reply]
So we can edit data, just not the schema?  Or we can only edit data in `users` and `identities` tables (or in auth schema) ?

**Reply by MossabDiae** [reply]
The above solution (adding user + identity) still works.
